Readme.md
#Nextboth Mobile Robot Simulator 2.0

Nextbot Robot Simulator implementation under GAZEBO

#How to install

* Be sure to have gazebo and all its indigo packages already installed
* Copy this package under your **catkin_ws/src/** workspace
* Edit **my_map2.yaml** and fix the path of **my_map2.pgm** relative to your laptop
* Edit **move_base.launch** and fix the path of **my_map2.yaml** relative to your laptop
* In the **catkin_ws** folder run **catkin_make clean**
* In the **catkin_ws** folder run **catkin_make --pkg nextbot_robot**
* In the **catkin_ws** folder run **catkin_make --pkg nextbot_teleop**
* You have to install the following packages
* *sudo apt-get install aptitude*
* *sudo apt-get install ros-indigo-joy*
* *sudo aptitude install ros-indigo-gazebo-ros-control*
* *sudo aptitude install ros-indigo-joint-state-controller*
* *sudo aptitude install ros-indigo-effort-controllers*
* *sudo aptitude install ros-indigo-laser-proc*
* *sudo aptitude install ros-indigo-gmapping*
* *sudo aptitude install ros-indigo-openslam-gmapping*

#How to build and save a map
* [IMPORTANT] Before following roslaunch commands, plz make sure the **ROS_IP** environment variable are set in all terminals
 * **export ROS_IP=x.x.x.x**   // x.x.x.x is the IP address of your ROS machine 
* Run **roslaunch nextbot_robot nextbot_simulator.launch** for the maze world and wait a bit, or
* Run **roslaunch nextbot_robot nextbot_simulator_apartment.launch** for the apartment world and wait a bit
* If you get errors like *Unable to set value [0,100000001] ...* stop the app, write *export LC_NUMERIC=C* and restart the launch
* Connect your joypad and run **roslaunch nextbot_teleop xbox360_teleop_nextbot.launch**
 * Push the right frontal button labeled as "RB" to activate cmd_vel publishing. Move the left stick around to control the velocity and the "A" button to rotate the laser scanner 
* If you dont have xbox360 joypad, you can use following node to teleop the robot by PC keyboard.

          $ rosrun nextbot_teleop nextbot_teleop_key1
          Arrow UP:    go forward
          Arrow DOWN:  go backward
          Arrow LEFT:  turn left
          Arrow RIGHT: turn right
          P key:       stop moving
          A:           start laser scan
          S:           stop laser scan

* Look at the **robot_mapping.launch** file, if you wanna use the single beam you have to set *arg name="scan_topic" default="scan"* , otherwise if you want the wide scan *arg name="scan_topic" default="laser"*
* Run **roslaunch nextbot_robot robot_mapping.launch** for the laser scan mapping
* Run **roslaunch nextbot_robot robot_mapping_sonar.launch** for the sonar scan mapping
* Run **rosrun nextbot_robot rotating_laser** for the laser scan mapping
* Run **rosrun nextbot_robot scan_from_sonar** for the sonar scan mapping
* If you are using the wide laser scans you can navigate within the room using the left stick of the joypad (remember to push the safety button on the top right and the A or X button of the XBOX joypad to rotate the laser scanner base)
* If you are using the single beam laser scans you have to stay motionless, get the 180° scan pushing the rotating base button, release the rotating base button of the joypad and only after that move (not so much) the robot to another point of the room to complete the other 180° scan of what is around it.
* After the full navigation you can save the map running **rosrun map_server map_saver -f /YOURCATKINWS/src/nextbot_simulator/nextbot_robot/NAMEOFYOURMAP**

#How to run the autonomous navigation

* [IMPORTANT] Before following roslaunch commands, plz make sure the **ROS_IP** environment variable are set in all terminals
 * **export ROS_IP=x.x.x.x**                  // x.x.x.x is the IP address of your ROS machine 
* Run **roslaunch nextbot_robot nextbot_simulator.launch**
* Run **roslaunch nextbot_robot move_base.launch**
* Go on Rviz, click on **2D Nav Goal** and point and click on the Rviz map as goal for the robot

<br><br><br>

All the files are edited or created by: <br>
Dott. Antonio Mauro Galiano<br>
**antoniomauro.galiano@gmail.com**<br>
Zach<br>
**qoogood1234@gmail.com**<br>
Kevin Kuei<br>
**kevin.kuei.0321@gmail.com**<br>
Only for simulation purpose<br>


nextbot_teleop/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package nextbot_teleop
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0 (2015-07-09)
-------------------



nextbot_teleop/src/nextbot_joy.cpp
/*
 *
 *  Created Or Edited on: 09/07/2015
 *      Author: Dott. Antonio Mauro Galiano
		antoniomauro.galiano@gmail.com
 */

#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <sensor_msgs/Joy.h>
#include "boost/thread/mutex.hpp"
#include "boost/thread/thread.hpp"
#include "ros/console.h"

class NextbotTeleop
{
public:
  NextbotTeleop();

private:
  void joyCallback(const sensor_msgs::Joy::ConstPtr& joy);
  void publish();

  ros::NodeHandle ph_, nh_;

  int linear_, angular_, deadman_axis_;
  double l_scale_, a_scale_;
  ros::Publisher vel_pub_;
  ros::Subscriber joy_sub_;

  geometry_msgs::Twist last_published_;
  boost::mutex publish_mutex_;
  bool deadman_pressed_;
  bool zero_twist_published_;
  ros::Timer timer_;

};

NextbotTeleop::NextbotTeleop():
  ph_("~"),
  linear_(1),
  angular_(0),
  deadman_axis_(4),
  l_scale_(0.3),
  a_scale_(0.9)
{
  ph_.param("axis_linear", linear_, linear_);
  ph_.param("axis_angular", angular_, angular_);
  ph_.param("axis_deadman", deadman_axis_, deadman_axis_);
  ph_.param("scale_angular", a_scale_, a_scale_);
  ph_.param("scale_linear", l_scale_, l_scale_);

  deadman_pressed_ = false;
  zero_twist_published_ = false;

  vel_pub_ = ph_.advertise<geometry_msgs::Twist>("cmd_vel", 1, true);
  joy_sub_ = nh_.subscribe<sensor_msgs::Joy>("joy", 10, &NextbotTeleop::joyCallback, this);

  timer_ = nh_.createTimer(ros::Duration(0.1), boost::bind(&NextbotTeleop::publish, this));
}

void NextbotTeleop::joyCallback(const sensor_msgs::Joy::ConstPtr& joy)
{ 
  geometry_msgs::Twist vel;
  vel.angular.z = a_scale_*joy->axes[angular_];
  vel.linear.x = l_scale_*joy->axes[linear_];
  last_published_ = vel;
  deadman_pressed_ = joy->buttons[deadman_axis_];
}

void NextbotTeleop::publish()
{
  boost::mutex::scoped_lock lock(publish_mutex_);

  if (deadman_pressed_)
  {
    vel_pub_.publish(last_published_);
    zero_twist_published_=false;
  }
  else if(!deadman_pressed_ && !zero_twist_published_)
  {
    vel_pub_.publish(*new geometry_msgs::Twist());
    zero_twist_published_=true;
  }
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "Nextbot_teleop");
  NextbotTeleop Nextbot_teleop;

  ros::spin();
}


nextbot_robot/src/scan_from_sonar.cpp
/*
 *
 *  Created Or Edited on: 12/10/2015
 *      Author: Dott. Antonio Mauro Galiano
 *		antoniomauro.galiano@gmail.com
 *
 *   This node creates the usefull scan from the sonar sensors to try to build a map
 */

#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <sensor_msgs/LaserScan.h>
#include <sensor_msgs/Range.h>
#include <geometry_msgs/Vector3.h>
#include <tf/transform_broadcaster.h>
#include <unistd.h>


double distance_scan;
double sonar_1, sonar_2, sonar_3, sonar_4, sonar_5, sonar_6, sonar_7, sonar_8, sonar_9, sonar_10, sonar_11, sonar_12; 
int rotate;
int count = 0;

void sonar1Callback(const sensor_msgs::Range& sonar1){
	sonar_1=sonar1.range;	
}

void sonar2Callback(const sensor_msgs::Range& sonar2){
	sonar_2=sonar2.range;	
}

void sonar3Callback(const sensor_msgs::Range& sonar3){
	sonar_3=sonar3.range;	
}

void sonar4Callback(const sensor_msgs::Range& sonar4){
	sonar_4=sonar4.range;	
}

void sonar5Callback(const sensor_msgs::Range& sonar5){
	sonar_5=sonar5.range;	
}

void sonar6Callback(const sensor_msgs::Range& sonar6){
	sonar_6=sonar6.range;	
}

void sonar7Callback(const sensor_msgs::Range& sonar7){
	sonar_7=sonar7.range;	
}

void sonar8Callback(const sensor_msgs::Range& sonar8){
	sonar_8=sonar8.range;	
}

void sonar9Callback(const sensor_msgs::Range& sonar9){
	sonar_9=sonar9.range;	
}

void sonar10Callback(const sensor_msgs::Range& sonar10){
	sonar_10=sonar10.range;	
}

void sonar11Callback(const sensor_msgs::Range& sonar11){
	sonar_11=sonar11.range;	
}

void sonar12Callback(const sensor_msgs::Range& sonar12){
	sonar_12=sonar12.range;	
}

int main(int argc, char** argv){
 	ros::init(argc, argv, "scan_from_sonar");
	ros::NodeHandle n;
	ros::Time current_time;
	
        std_msgs::Float64 laser_angle;
	unsigned int num_readings = 12;
  	double laser_frequency = 20;
  	double range[num_readings];
	sensor_msgs::LaserScan scan;
	scan.header.stamp = current_time;
	scan.header.frame_id = "sonar_1";
	scan.angle_min = -6.28;
	scan.angle_max = 0.0;
	scan.angle_increment = 6.28 / num_readings;
	scan.time_increment = (1 / laser_frequency) / (num_readings);
	scan.range_min = 0.02;
	scan.range_max = 4;
	scan.ranges.resize(num_readings);
	ros::Publisher sonar_laser_pub = n.advertise<sensor_msgs::LaserScan>("sonar_scan", 50);
	int flag = 0;

	ros::Subscriber sonar_1_subscriber_;
	sonar_1_subscriber_ = n.subscribe("nextbot/range_sonar_1", 100, sonar1Callback);
	ros::Subscriber sonar_2_subscriber_;
	sonar_2_subscriber_ = n.subscribe("nextbot/range_sonar_2", 100, sonar2Callback);
	ros::Subscriber sonar_3_subscriber_;
	sonar_3_subscriber_ = n.subscribe("nextbot/range_sonar_3", 100, sonar3Callback);
	ros::Subscriber sonar_4_subscriber_;
	sonar_4_subscriber_ = n.subscribe("nextbot/range_sonar_4", 100, sonar4Callback);
	ros::Subscriber sonar_5_subscriber_;
	sonar_5_subscriber_ = n.subscribe("nextbot/range_sonar_5", 100, sonar5Callback);
	ros::Subscriber sonar_6_subscriber_;
	sonar_6_subscriber_ = n.subscribe("nextbot/range_sonar_6", 100, sonar6Callback);
	ros::Subscriber sonar_7_subscriber_;
	sonar_7_subscriber_ = n.subscribe("nextbot/range_sonar_7", 100, sonar7Callback);
	ros::Subscriber sonar_8_subscriber_;
	sonar_8_subscriber_ = n.subscribe("nextbot/range_sonar_8", 100, sonar8Callback);
	ros::Subscriber sonar_9_subscriber_;
	sonar_9_subscriber_ = n.subscribe("nextbot/range_sonar_9", 100, sonar9Callback);
	ros::Subscriber sonar_10_subscriber_;
	sonar_10_subscriber_ = n.subscribe("nextbot/range_sonar_10", 100, sonar10Callback);
	ros::Subscriber sonar_11_subscriber_;
	sonar_11_subscriber_ = n.subscribe("nextbot/range_sonar_11", 100, sonar11Callback);
	ros::Subscriber sonar_12_subscriber_;
	sonar_12_subscriber_ = n.subscribe("nextbot/range_sonar_12", 100, sonar12Callback);

  	float angle;

        
  	ros::Rate r(50);
	
  	while(n.ok()){
			ros::Time scan_time = ros::Time::now();
			scan.header.stamp = scan_time;
			scan.ranges[0] = sonar_1;
			scan.ranges[1] = sonar_2;
			scan.ranges[2] = sonar_3;
			scan.ranges[3] = sonar_4;
			scan.ranges[4] = sonar_5;
			scan.ranges[5] = sonar_6;
			scan.ranges[6] = sonar_7;
			scan.ranges[7] = sonar_8;
			scan.ranges[8] = sonar_9;
			scan.ranges[9] = sonar_10;
			scan.ranges[10] = sonar_11;
			scan.ranges[11] = sonar_12;
			sonar_laser_pub.publish(scan);
			ros::spinOnce();
			r.sleep();
		}
  	
}


nextbot_teleop/src/nextbot_teleop_key1.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <signal.h>
#include <termios.h>
#include <stdio.h>
#include <sensor_msgs/Joy.h>

#define KEYCODE_R 0x43 
#define KEYCODE_L 0x44
#define KEYCODE_U 0x41
#define KEYCODE_D 0x42
#define KEYCODE_Q 0x71

#define KEYCODE_P 'p'   // for stop the robot
#define KEYCODE_A 'a'   // for starting laser scan
#define KEYCODE_S 's'   // for stopping laser scan

class TeleopTurtle
{
public:
  TeleopTurtle();
  void keyLoop();

private:

  
  ros::NodeHandle nh_;
  double linear_, angular_, l_scale_, a_scale_;
  ros::Publisher twist_pub_;
  ros::Publisher joy_pub_;
  
};

TeleopTurtle::TeleopTurtle():
  linear_(0),
  angular_(0),
  l_scale_(0.3),
  a_scale_(0.3)
{
  nh_.param("scale_angular", a_scale_, a_scale_);
  nh_.param("scale_linear", l_scale_, l_scale_);

  twist_pub_ = nh_.advertise<geometry_msgs::Twist>("nextbot/cmd_vel", 1);
  joy_pub_ =   nh_.advertise<sensor_msgs::Joy>("joy", 1);
}

int kfd = 0;
struct termios cooked, raw;

void quit(int sig)
{
  tcsetattr(kfd, TCSANOW, &cooked);
  ros::shutdown();
  exit(0);
}


int main(int argc, char** argv)
{
  ros::init(argc, argv, "teleop_turtle");
  TeleopTurtle teleop_turtle;

  signal(SIGINT,quit);

  teleop_turtle.keyLoop();
  
  return(0);
}


void TeleopTurtle::keyLoop()
{
  char c;
  bool dirty=false;
  bool dirty1=false;


  // get the console in raw mode                                                              
  tcgetattr(kfd, &cooked);
  memcpy(&raw, &cooked, sizeof(struct termios));
  raw.c_lflag &=~ (ICANON | ECHO);
  // Setting a new line, then end of file                         
  raw.c_cc[VEOL] = 1;
  raw.c_cc[VEOF] = 2;
  tcsetattr(kfd, TCSANOW, &raw);

  puts("Reading from keyboard");
  puts("---------------------------");
  puts("Use arrow keys to move the turtle.");


  for(;;)
  {
    // get the next event from the keyboard  
    if(read(kfd, &c, 1) < 0)
    {
      perror("read():");
      exit(-1);
    }

    linear_=angular_=0;
    ROS_DEBUG("value: 0x%02X\n", c);

    sensor_msgs::Joy joy;
    joy.buttons.resize(1);
    joy.buttons[0] = 0;
  
    switch(c)
    {
      case KEYCODE_L:
        ROS_INFO("LEFT");
        angular_ = 1.0;
        dirty = true;
        break;
      case KEYCODE_R:
        ROS_INFO("RIGHT");
        angular_ = -1.0;
        dirty = true;
        break;
      case KEYCODE_U:
        ROS_INFO("UP");
        linear_ = 1.0;
        dirty = true;
        break;
      case KEYCODE_D:
        ROS_INFO("DOWN");
        linear_ = -1.0;
        dirty = true;
        break;
      case KEYCODE_P:
        ROS_INFO("STOP MOVING");
        linear_ = 0.0;
        angular_ = 0.0;
        dirty = true;
        break;

      case KEYCODE_A:
        ROS_INFO("A");
        joy.buttons[0] = 1;
        joy_pub_.publish(joy);
        break;

      case KEYCODE_S:   // stop scanning
        ROS_INFO("S");
        joy.buttons[0] = 0;
        joy_pub_.publish(joy);
        break;

    }
   

    geometry_msgs::Twist twist;
    twist.angular.z = a_scale_*angular_;
    twist.linear.x = l_scale_*linear_;

    if(dirty ==true)
    {
      twist_pub_.publish(twist);    
      dirty=false;
    }

//    if (dirty1 == true)
//    {
//      joy_pub_.publish(joy);
//      dirty1=false;
//    }
  }


  return;
}



nextbot_robot/src/rotating_laser.cpp
/*
 *
 *  Created Or Edited on: xx/09/2015
 *      Author: Dott. Antonio Mauro Galiano
 *		antoniomauro.galiano@gmail.com
 *      Author: Kevin Kuei
 *		kevin.kuei.0321@gmail.com
 *   
 *   ATTENTION: "scan" topic is the one related to the single beam
 *              "laser" topic is the one related to the 180° laser message 
 */

#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <sensor_msgs/LaserScan.h>
#include <geometry_msgs/Vector3.h>
#include <sensor_msgs/Joy.h>
#include <tf/transform_broadcaster.h>
#include <unistd.h>


double distance_scan;
int rotate;
int count = 0;

void laserCallback(const sensor_msgs::LaserScan& laserscan){
	distance_scan = laserscan.ranges[1];	
}

void joyCallback(const sensor_msgs::JoyConstPtr& joy){
	rotate=joy->buttons[0];
	ROS_INFO("ROTATE %d",rotate);
}

int main(int argc, char** argv){
 	ros::init(argc, argv, "rotating_laser");
	ros::NodeHandle n;
	ros::Time current_time;
	
        std_msgs::Float64 laser_angle;
	unsigned int num_readings = 360;
  	double laser_frequency = 10;
  	double range[num_readings];
	sensor_msgs::LaserScan scan;
	scan.header.stamp = current_time;
	scan.header.frame_id = "laser";
	scan.angle_min = -3.14;
	scan.angle_max = 0.0;
	scan.angle_increment = 3.14 / num_readings;
	scan.time_increment = (1 / laser_frequency) / (num_readings);
	scan.range_min = 0.02;
	scan.range_max = 30.0;
	scan.ranges.resize(num_readings);
	ros::Publisher laser_pub = n.advertise<sensor_msgs::LaserScan>("laser", 50);
	int flag = 0;

  	
 	ros::Publisher laser_angle_pub = n.advertise<std_msgs::Float64>("/nextbot/laser_position_controller/command", 50);
	ros::Subscriber laser_subscriber_;
	laser_subscriber_ = n.subscribe("scan", 100, laserCallback);

	ros::Subscriber joy_subscriber_;
	joy_subscriber_ = n.subscribe("joy", 100, joyCallback);

  	float angle;

        angle = 0;

        ros::Rate r1(50);
        while(n.ok()) {
            angle -= 6.28/720;
            laser_angle.data = angle;
            laser_angle_pub.publish(laser_angle);
            
            r1.sleep();
            //ros::spinOnce(); 
            if (angle <= -1.57) break;
        }

        sleep(5); // to make sure the laser joint move to home position

        angle = -1.57;
  	ros::Rate r(50);
	//ROS_INFO("USCITO DA HOME STEP");
	//std_msgs::Float64 laser_angle;
  	while(n.ok()){
		if (rotate == 1){
			ROS_INFO("count %d",count);
			ROS_INFO("ANGLE %f",angle);

		        current_time = ros::Time::now();
		        laser_angle.data = angle;
			laser_angle_pub.publish(laser_angle);   // set laser joint position
		        //ros::spinOnce();

			scan.header.stamp = current_time;
			scan.ranges[count] = distance_scan;		

			if (count>=359) {
		        	laser_pub.publish(scan);  // publish Laserscan message
				while(n.ok()) {
				    angle -= 6.28/720;
				    laser_angle.data = angle;
				    laser_angle_pub.publish(laser_angle);
				    
				    r1.sleep();
				    //ros::spinOnce(); 
				    if (angle <= -1.57) break;
				}				
				//rotate=0;
				//
				angle = -1.57;
				count=0;
				sleep(5);
				ROS_INFO("count inside %d",count);
				ROS_INFO("ANGLE inside %f",angle);
		        }

		        if (count>=359) flag = 1;
		        else if (count<=0) flag = 0;
		        if (flag == 0){
		                angle  = angle + 0.008722222;
		                count++;
		        }
		        else if (flag == 1){
		                angle  = angle - 0.008722222;
		                count--;
		        }
		}
			ros::spinOnce();
			r.sleep();
		}
  	
}
